### ISSUE TITLE
Politely mention that `Box<&T>` is almost the most useless type you can have

### ISSUE THREAD

**Issue by oli-obk** [rust-lang-nursery/rust-clippy#2394]
Found in the wild: https://stackoverflow.com/questions/48406077/mutation-of-a-variable-inside-while-loop

List of types that make no sense that we should be linting about:

* `Box<&T>`
* `Box::new(SomeT)` where `sizeof::<T>() <= sizeof::<usize>()`
    * unless there are `Box::into_raw` calls within the function
* `Rc<Box<T>>`
* `Rc<Rc<T>>`
* `Rc<&T>`
* more...?

**Comment by clarcharr** [comment]
There's already a lint for `Box<Vec>`; why not extend this to the above types?

We don't include `Box<String>` iirc.

**Comment by killercup** [comment]
Hm, so `Box<&T>` is most likely by accident, I agree.

But thinking about the [box_vec](https://rust-lang-nursery.github.io/rust-clippy/v0.0.181/index.html#box_vec) lint I'm not sure I agree with its reasoning:

> Vec already keeps its contents in a separate area on the heap. So if you Box it, you just add another level of indirection without any benefit whatsoever.

There is a benefit I can think of: Boxing a Vec or String reduces the size from 3 words to 1. I'm sure some weird cases this may be a good idea, and, more importantly, this is most likely done _on purpose_. But maybe this is just a good idea in my imagination. Or maybe there is a crate that gives you a Vec-like API but stores length an capacity as part of the heap data and we should suggest using that?

**Comment by spacekookie** [comment]
Hi! I'd be interested in implementing this, the `Contributing.md` told me to leave a comment here and be assigned a mentor? 😄 

*comments*

**Comment by oli-obk** [comment]
👍 great

So a lot has already been mentioned here. Most importantly the fact that we already have related lints that could be extended to cover some of the cases here.

I think as a first (and uncontroversial) step you should concentrate on the `Box<&T>` case.

**Comment by killercup** [comment]
Not sure if it's mentioned in the Contributing file, _or_ how useful it is in this case, but clippy brings with it an `#[author]` attribute ([src](https://github.com/rust-lang-nursery/rust-clippy/blob/c85f651cfd8aed5ebd21007959602b6821df8f39/clippy_lints/src/utils/author.rs#L14-L46)) that generates pattern matching code for the items you added it to.

(Btw, saw your barrel crate earlier! Will definitely need to try it out! Oh, and: Have you seen [this stuff](https://github.com/diesel-rs/diesel/tree/c9bba07cc8ce6f2694122b8eca0d49205ae1c96a/diesel_tests/tests/schema_dsl)? :))

**Comment by spacekookie** [comment]
@killercup Hey 😄 Yea, I've seen those blocks in a few other lint files already. Although I'm not sure exactly how to start, at least from the ones I've looked at so far. Is there maybe one of the lints that especially easy so I can wrap my head around the structure?

(I talked to Sean yesterday about integrating barrel as an optional feature into Diesel 😄 Definitely excited about that!)

**Comment by Manishearth** [comment]
As an aside, I want to eventually change the author lint so that it lets us do coccinelle style lints, i.e. your lint code is just the code template you're looking for, wrapped in a macro.

**Comment by Manishearth** [comment]
Oh, and to answer the original question, I can mentor, but I'm out of town for another day and may have a spotty connection. But the tips in this issue should get you started, lmk if you have questions!

**Comment by spacekookie** [comment]
@Manishearth That's absolutely fine. I thought I'd drop the comment here tonight, didn't expect an immediate response 😅 

I'll be in a position to start work on this tomorrow. Should I drop you a message on gitter to stop spamming here then?

**Comment by Manishearth** [comment]
I'm not very active on gutter: spamming here is the best way to go. Or IRC, but IRC isn't good for asynchronous discussions.

If you're afraid of bothering folks, don't be. But if you want you can open a partial pull request and we can discuss these instead.

**Comment by spacekookie** [comment]
So I've finally gotten around trying to write some code for this but I have an issue. Commit 6f48e37d22b880389e5efe2e525650ff4da91681 doesn't actually compile for me which makes testing my own code rather difficult 😜 

Any way I can get around that?

**Comment by oli-obk** [comment]
Uh... what do you mean by "doesn't compile"? Are you on the most recent nightly?

**Comment by spacekookie** [comment]
Ah d'oh! I could have sword I ran `rustup update` before but I think I must have been on a nightly from a couple weeks ago

**Comment by darArch** [comment]
@spacekookie are you still working on this? It might provide a good first issue to tackle as part of the impl days, if you are working on it would you be able to update on any progress?

**Comment by spacekookie** [comment]
Hey, yea no I haven't done anything on this in a while >.> I still wanted to try to do it but if someone else wants to work on this, go ahead :slightly_smiling_face:

**Comment by clarcharr** [comment]
This isn't just a beginner issue; rust-lang/rust#51281 shows an example of this in the compiler. This should probably look through type aliases too.

**Comment by cauebs** [comment]
Is there something left to be done here?

Maybe this?
> - `Box::new(SomeT)` where `sizeof::<T>() <= sizeof::<usize>()`
>     - unless there are `Box::into_raw` calls within the function

I'm willing to work on it, just need to make sure it's wanted.

**Comment by jyn514** [comment]
> more...?

`Rc<String>` (should just be `Rc<str>`)

**Comment by Manishearth** [comment]
I'm unsure about small boxes, but it seems fine. Definitely worth making a PR

**Comment by joao-conde** [comment]
I this being worked on currently? Could I take it? Which remaining boxed types would we like to raise an alert for?

**Comment by HU90m** [comment]
```rs
fn foo(_: Box<&str>) {}
```

I've just noticed the above is already covered by `redundant_allocation`, which is nice.

**Comment by felix91gr** [comment]
For bookkeeping, I'm going to add here a reference to the `thin_box` issue. That way someone knows to ping this issue so that we can accommodate thin boxes within this lint later.

https://github.com/rust-lang/rust/issues/92791

**Comment by Coca162** [comment]
> `Rc<String>` (should just be `Rc<str>`)

`Rc<String>` could be useful if you wanted to use [try_unwrap](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.try_unwrap) since owning a `str` would not be possible
